src/inventory_ui.cpp

```diff
@@ -56,48 +129,260 @@ const item &inventory_entry::get_item() const {
 }
 
-inventory_entry *inventory_column::find_by_invlet( long invlet ) const
+inventory_selector_preset::inventory_selector_preset() : bug_text( _( "it's a bug!" ) )
 {
-    for( const auto &entry : entries ) {
+    append_cell( [ this ]( const inventory_entry & entry ) -> std::string {
+        return get_caption( entry );
+    } );
+}
+
+bool inventory_selector_preset::is_shown( const item & ) const
+{
+    return true;
+}
+
+bool inventory_selector_preset::is_shown( const item_location &location ) const
+{
+    return is_shown( *location.get_item() );
+}
+
+bool inventory_selector_preset::is_enabled( const item & ) const
+{
+    return true;
+}
+
+bool inventory_selector_preset::is_enabled( const item_location &location ) const
+{
+    return is_enabled( *location.get_item() );
+}
+
+int inventory_selector_preset::get_rank( const item & ) const
+{
+    return 0;
+}
+
+int inventory_selector_preset::get_rank( const item_location &location ) const
+{
+    return get_rank( *location.get_item() );
+}
+
+nc_color inventory_selector_preset::get_color( const item &it ) const
+{
+    return it.color_in_inventory();
+}
+
+nc_color inventory_selector_preset::get_color( const item_location &location ) const
+{
+    return get_color( *location.get_item() );
+}
+
+std::string inventory_selector_preset::get_caption( const inventory_entry &entry ) const
+{
+    const size_t count = entry.get_stack_size();
+    const std::string disp_name = entry.get_item().display_name( count );
+
+    return ( count > 1 ) ? string_format( "%d %s", count, disp_name.c_str() ) : disp_name;
+}
+
+std::string inventory_selector_preset::get_cell_text( const inventory_entry &entry,
+        size_t cell_index ) const
+{
+    if( cell_index >= cells.size() ) {
+        debugmsg( "Invalid cell index %d.", cell_index );
+        return bug_text;
+    }
+    if( entry.is_null() ) {
+        return std::string();
+    } else if( entry.is_item() ) {
+        return cells[cell_index].second( entry );
+    } else if( cell_index != 0 ) {
+        return cells[cell_index].first;
+    } else {
+        return entry.get_category_ptr()->name;
+    }
+}
+
+nc_color inventory_selector_preset::get_cell_color( const inventory_entry &entry, size_t ) const
+{
+    return entry.is_item() ? get_color( entry.get_location() ) : c_magenta;
+}
+
+size_t inventory_selector_preset::get_cell_width( const inventory_entry &entry,
+        size_t cell_index ) const
+{
+    return utf8_width( get_cell_text( entry, cell_index ), true );
+}
+
+void inventory_selector_preset::append_cell( const item_text_func &func, const std::string &title )
+{
+    append_cell( [ func ]( const inventory_entry & entry ) { // Don't pass by reference
+        return func( entry.get_item() );
+    }, title );
+}
+
+void inventory_selector_preset::append_cell( const item_location_text_func &func,
+        const std::string &title )
+{
+    append_cell( [ func ]( const inventory_entry & entry ) { // Don't pass by reference
```

Comment:
I meant "Don't capture by reference", yes. The `func` isn't guaranteed (nor needed) to live longer than the class.